### ISSUE TITLE
Provide option to generate client certificates with ECDSA keys

### ISSUE THREAD

**Issue by arikalfus** [pivpn/pivpn#53]
Elliptic curve cryptography has been fully supported in OpenVPN since 2.4.0 (~2014) and is much less computationally intensive than RSA - ideal for client certificates going on mobile phones. There was a commit in 2013 in the easy-RSA changelog adding ECDSA support, so it's already built-in there too. Hopefully, this option can be provided for pivpn as well.

**Comment by 0-kaladin** [comment]
not sure what exactly would have to be done.  I'd have to research.  already set to tls min 1.2 so you surely aren't using any weak ciphers at this point.

**Comment by 0-kaladin** [comment]
have to see what this entails.  the protocol already set to tls min v 1.2 so no bad ciphers are being used.  I'd have to research how to get easy rsa to use a specific cipher.  and I don't want it to break any clients either.

**Comment by redfast00** [comment]
[This isn't supported in Easy-RSA 2.0](https://github.com/OpenVPN/openvpn/blob/master/README.ec). :(

**Comment by arikalfus** [comment]
Copying relevant paragraph from above link, yeah requires Easy-RSA 3.0:

"Support for generating an ECDSA certificate chain is available in EasyRSA (in
spite of it's name) since EasyRSA 3.0. The parameters you're looking for are
'--use-algo=ec' and '--curve=<curve_name>'. See the EasyRSA documentation for
more details on generating ECDSA certificates."

**Comment by 0-kaladin** [comment]
hmm so should this be an option or if implemented simply used?  I don't want to present too many options especially why would someone not want to use this?

**Comment by 0-kaladin** [comment]
@arikalfus I started this.  It won't be an option, we'll just do it.
We'll use our own pivpn releases of easy-rsa to accomplish:
https://github.com/pivpn/easy-rsa/releases/tag/3.0.1-pivpn

**Comment by philgithub1** [comment]
So the auto update will cover this in existing installs? Where does one check if it has been updated?

**Comment by arikalfus** [comment]
Sorry the the delay in responding. Re: why would someone be want an ECDSA key - can't think of a reason.

Cloudflare has a good article from a few years ago on why ECDSA keys are 'better' than RSA keys. https://blog.cloudflare.com/ecdsa-the-digital-signature-algorithm-of-a-better-internet/

TLDR quote: "Why is ECDSA the algorithm of choice for new protocols when RSA is available and has been the gold standard for asymmetric cryptography since 1977? It boils down to the fact that we are better at breaking RSA than we are at breaking ECC."

An ECC-256 key is generally equivalent to an RSA 2048-bit key, and an ECC-512 bit key is equivalent to RSS 4096-bits. If you're going to choose just one, go with ECC-512 I'd say.

**Comment by redfast00** [comment]
@0-kaladin: Are you currently working on this? If not, can I give it a try?

**Comment by 0-kaladin** [comment]
@redfast00 it was next on my list to do over next couple days but I don't mind deferring to you if you want to take a stab at it.  
The only thing i did so far was add this var in the install.sh
easyrsaRel="https://github.com/pivpn/easy-rsa/releases/download/3.0.1-pivpn/EasyRSA-3.0.1-pivpn.tgz"

basically, don't install easy-rsa from apt.  instead wget that.  extract to easy-rsa then should have the ECDSA options.

I just updated the test branch with master if you want to work from there.

**Comment by redfast00** [comment]
@0-kaladin: what editor are you using? Every time you change a line, there is a trailing white space, and this results in weird diff's when comparing code.

**Comment by 0-kaladin** [comment]
Vi. Or vim .  I can ensure that doesn't happen in the future.

**Comment by redfast00** [comment]
Don't worry too much about it. (BTW, you can use `:%s/\s\+$//e` to [strip all trailing whitespaces in vim](http://vim.wikia.com/wiki/Remove_unwanted_spaces).)

**Comment by redfast00** [comment]
I can't get easyrsa3 to generate ECDSA keys. I have set the env variables `$EASYRSA_ALGO` and `$EASYRSA_CURVE` to respectively `ec` and `secp256r1`. Both with the `./easyrsa build-ca` and the `./easyrsa build-client-full`, it says it's using RSA 2048 bits. 
This will also be a lot of effort to rewrite the install script and the pivpn script. Also, [mandatory XKCD](https://xkcd.com/538/). It would be cool to have ECDSA, but I don't really see the added value for home users and this might introduce new bugs. I don't know if ECDSA is supported on all clients we currently support.

**Comment by 0-kaladin** [comment]
Let me take a look and see what is involved.

**Comment by arikalfus** [comment]
The value is in the reduced computational intensity (for me, this allows me to run a VPN on my mobile phone sacrificing much less battery power). However, the point about supporting all client is a good one - I'm not sure what the cutoff is, but there are some "old" clients that won't support it. That was the reason behind my original request for it to be an added option, as opposed to replacing the keys. Whether the power/speed savings are worth it vs. the necessary refactoring I'd have to leave to you guys to determine.

Here's an article comparing the speed and computational efficiency of RSA vs ECC: https://securitypitfalls.wordpress.com/2014/10/06/rsa-and-ecdsa-performance/

Summary: ECDHE and  ECDSA with 256 bit curves is 2.7 times faster than 2048 bit RSA with 256 bit ECDHE and 3.4 times faster than 3072 bit RSA alone.

**Comment by 0-kaladin** [comment]
I got it going:
```
root@raspberrypi:/etc/openvpn/easy-rsa/easyrsa3# ./easyrsa build-ca

Note: using Easy-RSA configuration from: ./vars
Generating a 384 bit EC private key
writing new private key to '/pki/private/ca.key.eoNTzv7aeX'
Enter PEM pass phrase:
Verifying - Enter PEM pass phrase:
```

let me play around with it a bit, might not be too hard.  easy-rsa3 seems to simplify a few things.
I'm not sure clients will be an issue, Note that i have tls 1.2 as a parameter already for clients which no old clients would support.  
I suspect anyone using this is installing on their android, iphone, windows, linux, or mac.  All have modern clients supporting the latest.  I just don't see old clients coming into play here but it'd be revealed in my testing if any of the current ones break.

**Comment by 0-kaladin** [comment]
I haven't had a ton of free time.  just pushed first set of changes up to test branch.  
that covers the core server changes.  next up server.conf adjustments then the pivpn side for clients.

**Comment by 0-kaladin** [comment]
@arikalfus @redfast00 
almost done.  just working on pivpn add, pivpn list (done), pivpn revoke now. 
Will you guys have any time to test before I push to master?  I've been unit testing a ton but the more the merrier.

**Comment by redfast00** [comment]
I'm preparing for my exams and won't have any time to test it out. Will definitely check it out after they're done.

**Comment by arikalfus** [comment]
@0-kaladin I could test this when it's ready. On Raspbian and Ubuntu, which I don't have a doubt you're covering, but for an extra 👍 that it's working.

**Comment by 0-kaladin** [comment]
It is ready from the testing branch now. Wiki has the curl for getting from
that branch. Thanks.
I still have more testing myself before I push to master.

**Comment by 0-kaladin** [comment]
## Short Story
EC needs openvpn 2.4 on the server, which we can't get reliably install + it is only RC1 at this time.  I've merged all my modifications anyway as easyrsa3 is better, easier, and now we can trivially go to EC if/when it is prudent.

## Long Story
So all the changes I've merged made a ton of little improvements, but of course the effort was switching everything to use easyrsa3.  This (unlike openvpn 2.4) has been out for a while and doesn't get frequent updates to source.  So I have a fork of it here in the pivpn org where i can merge any changes they make and cut our own releases from it.  This is a good setup for easy-rsa due to what I mentioned above but not something I'd like to do for openvpn itself.  I'd prefer to get that from a deb or package etc and there is no openvpn 2.4 I see for the arm architecture.    
The benefits of merging everything except the EC piece (I moved that back to RSA), is that with easyrsa3 there is a cn-only mode.  IE we don't need to supply country, org etc all the stuff that doesn't matter in these types of certificates.  You also don't source vars it just reads the file with options.  Generally just using easyrsa3 is a bit easier and cleaner IMO.  
I'll close this issue and we can continue to discuss EC with openvpn 2.4 in the new issue I created that is referenced above.
